Add preview options for update target list #72

kunif opened the pull request:
Add following options
  --preview, -p
  --preview-only, -po

display by pip list --outdated output format, before execution

kunif committed:
Add preview options for update target list

Add following options
  --preview, -p
  --preview-only, -po

display by pip list --outdated output format, before execution

jgonggrijp reviewed (changes requested):
Thank you for going out of your way to make a contribution to pip-review!

I like the functionality that you are aiming to add, but the quality needs to improve. In particular:

- [x] The concern of formatting a table should be separated from the other concerns (i.e., extracting the data to be formatted from the `outdated` list and logging the formatted table). I want you to write a function `format_table` that takes a tuple or list of equal-length columns, each being an iterable containing just the data that need to go into that column. Column headers can be passed in a second tuple with the same length as the first argument, or they can be the first element of each column (your choice). There must be no assumptions about the number of columns or their meanings. The function should return the formatted table as a string, without printing or logging anything. I can give you some hints on this if desired.
- [x] I also want you to make your code DRY (meaning "Don't Repeat Yourself"). Again, I can give hints if desired.
- [ ] Ideally, you should add a test to ensure that the table is formatted correctly. Testing is a bit difficult with our current setup, so this is not a hard requirement.

See my comments below for a few more details. Please feel welcome to ask all the questions you may have, we can take this ride together.

kunif commented on README.rst:
```diff
@@ -39,6 +39,36 @@ Example, run interactively, ask to upgrade for each package:
+
+    $ pip-review --preview-only
+    Package  Version Latest Type
+    -------- ------- ------ -----
+    redis    2.4.9   2.6.2  wheel
+    requests 0.13.2  0.14.0 wheel
+    rq       0.3.0   0.3.4  wheel
+    -----------------------------
```
I thought that the interrupted line at the top would be the same as the display of `pip list --outdated`.
A continuous line at the bottom marks the end of the list and is included to distinguish it from the --interactive and --auto display.
I want to keep the line that shows the delimiter, so make both lines continuous.

kunif commented on pip_review/__main__.py:
```diff
@@ -80,6 +80,12 @@ def parse_args():
+        '--preview', '-p', action='store_true', default=False,
+        help='Preview update target list before execution')
+    parser.add_argument(
+        '--preview-only', '-po', action='store_true', default=False,
```
This changes that way.

kunif commented on pip_review/__main__.py:
```diff
@@ -211,6 +217,19 @@ def get_outdated_packages(forwarded):
 
 
+def get_values_maxlength(outdated):
+    temp = (sub['name'] for sub in outdated)
+    name = max(len(elem) for elem in temp if elem is not None)
+    name = max([name, 7])
+    temp = (sub['version'] for sub in outdated)
+    version = max(len(elem) for elem in temp if elem is not None)
+    version = max([version, 7])
+    temp = (sub['latest_version'] for sub in outdated)
+    latest = max(len(elem) for elem in temp if elem is not None)
+    latest = max([latest, 6])
```
Along with the comment above, I'm sorry for that.
Now that I have come up with a more concise process, I will replace it with the following.
```
    namelen = max(len('Package'), *(len(e['name']) for e in outdated))
    versionlen = max(len('Version'), *(len(e['version']) for e in outdated))
    latestlen = max(len('Latest'), *(len(e['latest_version']) for e in outdated))
```

jgonggrijp replied:
That is *still* repeating yourself! Also, it is still mixing the concern of extracting the data from `outdated` with the concern of formatting a table.

kunif replied:
I don't understand what the point was.

What I'm doing here is finding the maximum length of the string that each column contains.
Depending on the contents of the outdated packages, they have different lengths and cannot be processed with just a predetermined length.

It looks like I'm extracting the data here because otherwise I can't get the string length of each row in the specified column.
Using the maximum string length of each column obtained in this way, the layout of the table is created so that it looks neat.

After that, a table is created based on the data of each package.

Is it better to judge the length at the same time as making the table?
I feel that such processing is more complicated and easier for bugs to enter.

jgonggrijp replied:
Look at this.

```py
from operator import itemgetter

# Single source of truth about columns. The remainder of the code
# makes no assumptions about names, order or number of columns.
COLUMNS = {
    'Package': 'name',
    'Version': 'version',
    'Latest': 'latest_version',
    'Type': 'latest_filetype',
}

# Next two functions describe how to collect data for the
# table. Note how they are not concerned with columns widths.

def extract_column(data, field, title):
    return [title] + list(map(itemgetter(field), data))

def extract_table(outdated):
    return [
        extract_column(outdated, field, title)
        for title, field in COLUMNS.items()
    ]

# Next two functions describe how to format any table. Note that
# they make no assumptions about where the data come from.

def column_width(column):
    return max(map(len, filter(None, column)))

def format_table(columns):
    widths = list(map(column_width, columns))
    row_fmt = ' '.join(map('{{:<{}}}'.format, widths)).format
    ruler = '-' * (sum(widths) + len(widths) - 1)
    rows = list(map(row_fmt, *columns))
    head = rows[0]
    body = rows[1:]
    return '\n'.join([head, ruler] + body + [ruler])

# Clean separation of concerns, no repetition!

def main():
    # ...
    logger.info(format_table(extract_table(outdated)))
```

kunif commented on pip_review/__main__.py:
```diff
@@ -223,11 +242,23 @@ def main():
             logger.info('{0}=={1}'.format(pkg['name'], pkg['latest_version']))
+        return
+    elif args.preview or args.preview_only:
+        namelen, versionlen, latestlen = get_values_maxlength(outdated)
+        previewformat = '{0:<' + str(namelen) + '} {1:<' + str(versionlen) + '} {2:<' + str(latestlen) + '} {3}'
+        logger.info(previewformat.format('Package','Version','Latest','Type'))
+        logger.info('-' * namelen + ' ' + '-' * versionlen + ' ' + '-' * latestlen + ' -----')
+        for pkg in outdated:
+            logger.info(previewformat.format(pkg['name'], pkg['version'], pkg['latest_version'], pkg['latest_filetype']))
+        logger.info('-' * (namelen + versionlen + latestlen + 8))
```
I did this to make the `logger` object belong to `main` and to make it similar to `--raw` and `--interactive` processing.

As in the summary of the first point, this defines a function that creates the whole to be displayed in preview as one string containing multiple lines, and in the processing of this `main` part, the result of that function is displayed in `logger`. Then it will be simple.

jgonggrijp replied:
I'm fine with there being a function called `format_preview`, which does both data extraction and table formatting. Still, I want you to *also* write a `format_table` function which does table formatting *only*. `format_preview` should then call `format_table` as the last step.

I'm OK with you skipping the tests, but that makes proper factoring even more important.

kunif replied:
I see.
It is possible to try to understand if it is specifically presented.

However, I didn't have this idea, and I could only write code with simple steps.
Well, it can be said that there is a lack of knowledge and experience and cultural differences.

By the way, as a result, you have recoded everything.
So this pull request is just like being rejected.
Please handle it that way.

kunif commented on pip_review/__main__.py:
```diff
@@ -223,11 +242,23 @@ def main():
+        if args.preview_only:
+            return
+    if args.auto:
+        update_packages(outdated, install_args)
```
In the comment above, change `elif` to` if`, and deal with it as pointed out.

kunif committed:
Reflection of review results

- Changed the line that distinguishes the item name and contents of the displayed table so that there is no break.
- Changed short form of --preview-only option from -po to -P.
- Replaced the process of formatting the table contents to be displayed in the review results.
- By changing the branch of main processing with if elif else to if and return within it, the --interactive option processing is extracted from else while keeping the execution judgment and indentation level clean.

- Incorporation of master updates.

jgonggrijp commented:
@kunif I see you updated this branch. Please clarify your intentions; are you ready for a new review?

CC @someparsa

kunif commented:
I'm not familiar with how to use github, so I don't know what to do next, but in the second commit, all your review results are reflected, and the updated contents of master and develop are also included.
Please let me know if there is anything I should do.

jgonggrijp commented on pip_review/__main__.py:
```diff
@@ -211,6 +235,45 @@ def get_outdated_packages(forwarded):
 
 
+# Single source of truth about columns. The remainder of the code
+# makes no assumptions about names, order or number of columns.
```
Ah, I wrote these comments specifically for you, to clarify what I meant. They don't need to be in `__main__.py`. You coud add a shorter comment for future editors just to explain the purpose of the variable:

```suggestion
# Nicer headings for the columns in the oudated package table.
```

Or something like that.

jgonggrijp commented on pip_review/__main__.py:
```diff
@@ -211,6 +235,45 @@ def get_outdated_packages(forwarded):
+    body = rows[1:]
+    return '\n'.join([head, ruler] + body + [ruler])
+
+# Clean separation of concerns, no repetition!
```
This comment should be removed.

jgonggrijp reviewed (commented):
> I'm not familiar with how to use github, so I don't know what to do next, but in the second commit, all your review results are reflected, and the updated contents of master and develop are also included.
> Please let me know if there is anything I should do.

This is what I wanted to know: you are done processing my review comments. Generally when the reviewer requests changes, you can follow up and then just post a new comment to notify the reviewer to look again.

There is one comment I would like you to change and one that I think should be removed; see detailed comments below. Other than that, I think this is ready to merge!

kunif committed:
Reflection of second review results

Fixed some comments

kunif commented:
Reflected the corrections pointed out in the second review.
Please confirm that.

jgonggrijp commented:
Confirmed! Thanks for contributing. Could you address the merge conflicts? Otherwise I'll do it sometime soon.

kunif commented:
Excuse me.
I'm not familiar with GitHub's detailed processing and operation procedures, so I'm not confident in resolving conflicts.
Could you please take care of it?

jgonggrijp commented:
Alright, will do!

someparsa commented:
Hi Kunif @kunif, I had a look over GIT guides online. Resolving conflicts is possible via both the terminal and the website. You can just open the pull through the web browser, see the conflicts, comment on them, accept them or just simply ignore them. I assume it should be very easy. I learnt so much by just an hour of online reading and watching.

jgonggrijp commented:
@someparsa It is a bit more complicated than it seems at first sight. Also, it is no problem for me to do it, I just need a quiet moment.

jgonggrijp commented:
Merged. Congratulations, @kunif!
